### ISSUE TITLE
Prepare blue-print for DA discussion & Contribution towards Reference Architecture Settlement efforts

### ISSUE THREAD

**Issue by PaulGregoryBaker** [mojaloop/project#3061]
## **Goal**:
`As a` member of the DA committee 
`I want to` understand clearly what is being asked to review
`so that` I can approve and support the effort to deliver the first steps of the vNext settlement
`by` allocating transfers to settlement models, batches based on metadata in transfers themselves (such as using SubScenario field)

**Acceptance Criteria**:
- [ ] Documentation of the proposed design that is compatible with ref. architecture (/vNext)
- [ ] Sequence diagrams depicting the proposed settlement model are documented

**Complexity**: <High|Medium|Low> > A short comment to remind the reason for the rating

**Uncertainty**: <High|Medium|Low> > A short comment to remind the reason for the rating

**Notes**:
- Overview of changes needed for the MAS PoC
- Characteristics of the Rules engine and where it goes
- Section detailing reference architecture compatibility

---

**Tasks**:
- [ ] TBD [ @? ]

**Done**
- [ ] Acceptance Criteria pass
- [ ] Designs are up-to date
- [ ] Unit Tests pass
- [ ] Integration Tests pass
- [ ] Code Style & Coverage meets standards
- [ ] Changes made to config (default.json) are broadcast to team and follow-up tasks added to update helm charts and other deployment config.
- [ ] TBD

## **Pull Requests**:
- [ ] TBD

## **Follow-up**:
- N/A

**Dependencies**:
- N/A

## **Accountability**:
- Owner: TBC
- QA/Review: TBC

**Comment by elnyry-sam-k** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/mojaloop-project-59edee71d1407922110cf083/issues/mojaloop/project/3061?planning-poker) with Zenhub @PaulGregoryBaker @mdebarros @kleyow @vijayg10

**Comment by PaulGregoryBaker** [comment]
# Implementation of transactional rules engine for determining Settlement Model
The scope of this effort is limited to the implementation and characterization of the transaction based rules engine that provides the transaction object and then applies a rule to determine the settlement model. This settlement model is then used to lookup both participants position and liquidity accounts.

## Settlement Reference Architecture Implementation Alignment
The next version of the Mojaloop Settlement design as outlined in the Reference Architecture; requires a rule to be configured to determine the settlement model based on the properties of a transfer. This work effort builds that rules engine and this is the contributed alignment to the reference architecture. In the reference architecture documentation this is referred to as 'Settlement BC running in the Transaction BC' in the transfer bounded context. This functionality will additionally be incorporated into the current Mojaloop version and built into the Prepare Handler.

Reference Architecture Alignment Contribution:
1. **Rules engine implementation** that is aligned with the vNext Settlement implementation and the Reference Architecture; with the associated interfaces defined. Steps 5 & 6 in the sequence diagram.
2. **Performance Characterization of the rules engine** when compared to normal javascript.

### Implementation Sequence Diagram in Current Mojaloop Version
![MASImplementation](https://user-images.githubusercontent.com/40144680/211555589-c685121a-380f-4e22-a1b3-eb4fce718d4e.svg)

## Incorporating functionality into current Mojaloop version

Here is a list of changed that are to be made in order to build this functionality into the current Mojaloop Version

1. Unpacking of the ILP in the ML API Adapter, and making the ILP data packet details available through the transfer object. (This data must be made available to the rules engine, but must only happen once to avoid multiple decodes of the ILP packet.) Steps 2-4 in the sequence diagram. 
2. Hosting of the rules engine in the Prepare handler to calculate the settlement account type. Steps 5-6 in the sequence diagram. 
3. Modification of how the participants settlement accounts and position accounts are identified. Step 7 in the sequence diagram.  
4. Ensure that the correct liquidity checks, limit checks and fund reservations are made. There will be not changed made to how these checks are made, only that the accounts against which these checks are made is based on the settlement account type provided by the rules engine. Step 8 in the sequence diagram. 

## Performance impact of implementation
There are two potential places where performance is impacted by this proposed change. All of these need to be during the processing of a transaction.
1. **Unpacking the ILP packet**
   
   The information required by the rule-engine is stored within this object, so this is a necessary and an unavoidable change. This change is being made in the ML API Adapter, so it can be horizontally scaled and subsequently doesn't affect the critical path only the transaction latency.
1. **Running the rules engine**

   This is a requirement in order to align this work to the reference architecture design. Effort is being made in this project to characterize the performance impact of this, and will be specified and delivered with this piece of work.

## Detailed view of Reference architecture alignment
### Reference Architecture Transfer Sequence Diagram
The red rectangle highlights where contributions are being made to the reference architecture.

![ReferenceArchDiag](https://user-images.githubusercontent.com/40144680/211555663-c348a7d8-2164-4d07-9005-aaf793b0f685.png)

### Items that will not be built as part of this work effort
1. Bootstrapping of the Settlement BS running the the transfer BC (using the platform configuration BC and Settlement BC). Instead the bootstrapping method will be aligned with the current solution bootstrapping.
2. Calling the participant BC in the running of the transaction BC is not necessary in the current Mojaloop solution, so this will not be built.
3. Accounts BC and Participant Lifecycle BC as they do not exist in the current Mojaloop solution, will not be notified.
4. The process of creating a participant, creating a participant account, depositing and withdrawing funds, and settling between participants will remain unchanged.

**Comment by PaulGregoryBaker** [comment]
Update the sequence diagram to reflect implementation changes that became apparent during the implementation.

**Comment by pedrosousabarreto** [comment]
As discussed at the DA meetings of the [7th](https://community.mojaloop.io/t/da-meeting-minutes-2022-12-07/470) and [11th](https://community.mojaloop.io/t/da-meeting-minutes-2023-01-11/472) of December of 2023 That red rectangle is not correct and was decided to be simple call to a library that returns the settlement model without any network request.
Rational was, and this was agreed at the DA, that we cannot have a intermediate step that calls another component over the network for every transfer prepare. This avoidance is critical for performance.
In summary, that red rectangle no longer exists and the reference architecture must be updated.

Additionally, this was not what was agreed at the DA and there was no formal review of this issue nor formal agreement.
One of the resulting actions of the DA meeting were to document the settlements requirements; those requirements are not yet available to the date of this comment.

I really don't understand why all of the activity in this issue happened after those DA sessions given the discussions that were had in said sessions.

**Comment by PaulGregoryBaker** [comment]
Hi Pedro,
Thanks for your comment.
A couple of points.
1. This work does not need DA approval, however it does need to be show and discussed with the DA. I.e. it is work being done that aligns with the reference architecture so should be shared with everyone. It is up to each team to subsequently take advantage of that work effort if it is possible to do so.
2. The Reference Architecture is what is documented. I.e. [here](https://mojaloop.github.io/reference-architecture-doc/boundedContexts/transfers/#use-cases). I understand that there is a process to change aspects of the reference architecture, and I have made concerted effort that the work being done would also align to those changes. (Do you know why the changes to the reference architecture have not yet been made?)  I am happy to come back and update this diagram once the change has been made. (I also am unable to make the change as I do not have the correct access to the diagrams.)
3. I disagree with your conclusion on the 7th and 11th of December 2022. Please refer to the meeting notes.
